src/Content/OEmbed.php

```diff
@@ -48,132 +47,128 @@ public static function replaceCallback($matches)
 		// left here as a fallback in case this is called from another source
-
-		$noexts = ["mp3", "mp4", "ogg", "ogv", "oga", "ogm", "webm"];
+		$noexts = ['mp3', 'mp4', 'ogg', 'ogv', 'oga', 'ogm', 'webm'];
 		$ext = pathinfo(strtolower($embedurl), PATHINFO_EXTENSION);
 
+		$oembed = new \Friendica\Object\OEmbed($embedurl);
 
-		if (is_null($txt)) {
-			$txt = "";
+		if ($json_string) {
+			$oembed->parseJSON($json_string);
+		} else {
+			$json_string = '';
 
 			if (!in_array($ext, $noexts)) {
 				// try oembed autodiscovery
 				$redirects = 0;
-				$html_text = Network::fetchUrl($embedurl, false, $redirects, 15, "text/*");
+				$html_text = Network::fetchUrl($embedurl, false, $redirects, 15, 'text/*');
 				if ($html_text) {
 					$dom = @DOMDocument::loadHTML($html_text);
 					if ($dom) {
 						$xpath = new DOMXPath($dom);
 						$entries = $xpath->query("//link[@type='application/json+oembed']");
 						foreach ($entries as $e) {
-							$href = $e->getAttributeNode("href")->nodeValue;
-							$txt = Network::fetchUrl($href . '&maxwidth=' . $a->videowidth);
+							$href = $e->getAttributeNode('href')->nodeValue;
+							$json_string = Network::fetchUrl($href . '&maxwidth=' . $a->videowidth);
 							break;
 						}
+
 						$entries = $xpath->query("//link[@type='text/json+oembed']");
 						foreach ($entries as $e) {
-							$href = $e->getAttributeNode("href")->nodeValue;
-							$txt = Network::fetchUrl($href . '&maxwidth=' . $a->videowidth);
+							$href = $e->getAttributeNode('href')->nodeValue;
+							$json_string = Network::fetchUrl($href . '&maxwidth=' . $a->videowidth);
 							break;
 						}
 					}
 				}
 			}
 
-			$txt = trim($txt);
-
-			if (!$txt || $txt[0] != "{") {
-				$txt = '{"type":"error"}';
-			} else { //save in cache
-				$j = json_decode($txt);
-				if (!empty($j->type) && $j->type != "error") {
-					DBA::insert('oembed', [
-						'url' => normalise_link($embedurl),
-						'maxwidth' => $a->videowidth,
-						'content' => $txt,
-						'created' => DateTimeFormat::utcNow()
-					], true);
-				}
+			$json_string = trim($json_string);
 
-				Cache::set($a->videowidth . $embedurl, $txt, CACHE_DAY);
+			if (!$json_string || $json_string[0] != '{') {
+				$json_string = '{"type":"error"}';
 			}
-		}
 
-		$j = json_decode($txt);
+			$oembed->parseJSON($json_string);
+			if (!empty($oembed->type) && $oembed->type != 'error') {
+				DBA::insert('oembed', [
+					'url' => normalise_link($embedurl),
+					'maxwidth' => $a->videowidth,
+					'content' => $json_string,
+					'created' => DateTimeFormat::utcNow()
+				], true);
+			}
 
-		if (!is_object($j)) {
-			return false;
+			Cache::set($cache_key, $json_string, CACHE_DAY);
 		}
 
-		// Always embed the SSL version
-		if (isset($j->html)) {
-			$j->html = str_replace(["http://www.youtube.com/", "http://player.vimeo.com/"], ["https://www.youtube.com/", "https://player.vimeo.com/"], $j->html);
+		if ($oembed->type == 'error') {
+			return $oembed;
```

Comment:
Why do we return now the variable ```$oembed``` instead of a ```false```?

Reply:
Because the functions calling this expect an OEmbed object, not a boolean.